File: day3/code.py
def getnum(s, pos):
    nums = []
    i = 0
    while i < len(s):
        for j in reversed(range(i, len(s))):
            if s[i:j + 1].isdigit():
                nums += [(int(s[i:j + 1]), i, j, pos)]
                i = j
        i += 1 
    return nums


lines = open("input.txt").readlines()
nums = []
n, m = len(lines), len(lines[1]) - 1
for i in range(n):
    nums += getnum(lines[i], i)

sym = []
for i in range(n):
    for j in range(m):
        if lines[i][j] != '.' and not lines[i][j].isdigit():
            sym += [(i, j)]

sum = 0
for (val, l, r, x) in nums:
    f = False 
    for i in range(max(0, x - 1), min(n, x + 2)):
        for j in range(max(0, l - 1), min(m, r + 2)):
            f |= ((i, j) in sym)
    if f:
        sum += val 
print("task1:", sum)

sum = 0
for (x, y) in sym:
    num = []
    for i in range(max(0, x - 1), min(n, x + 2)):
        for j in range(max(0, y - 1), min(m, y + 2)):
            for (val, l, r, xx) in nums:
                if xx == i and l <= j and j <= r:
                    num += [(val, l, r, xx)]
    num = list(set(num))
    if len(num) == 2:
        sum += num[0][0] * num[1][0]
print("task2:", sum)

File: day3/code.cpp
#include <bits/stdc++.h>
using namespace std;

void task1(vector<pair<int, tuple<int, int, int>>> nums, vector<vector<char>> sym) {
    int sum = 0;
    for (auto [val, pos] : nums) {
        auto [x, y, len] = pos;
        bool f = false;
        for (int i = max(0, x - 1); i < min((int)sym.size(), x + 2); ++i)
            for (int j = max(0, y - len); j < min((int)sym[0].size(), y + 2); ++j) 
                f |= sym[i][j];
        if (f)
            sum += val;
    }
    cout << "task1: " << sum << endl;
}

void task2(vector<pair<int, tuple<int, int, int>>> nums, vector<vector<char>> sym) {
    vector<vector<vector<int>>> b(sym.size(), vector<vector<int>>(sym[0].size()));
    for (auto [val, pos] : nums) {
        auto [x, y, len] = pos;
        bool f = false;
        for (int i = max(0, x - 1); i < min((int)sym.size(), x + 2); ++i)
            for (int j = max(0, y - len); j < min((int)sym[0].size(), y + 2); ++j) 
                if (sym[i][j] == '*')
                    b[i][j].push_back(val);
    }
    long long sum = 0;
    for (int i = 0; i < sym.size(); ++i)
        for (int j = 0; j < sym[0].size(); ++j)
            if (b[i][j].size() == 2)
                sum += 1ll * b[i][j][0] * b[i][j][1];
    cout << "task2: " << sum << endl;
}


int main() {
    freopen("input.txt", "r", stdin);
    vector<string> a;
    a.emplace_back();
    while (getline(cin, a.back()))
        a.emplace_back();
    a.pop_back();
    vector<pair<int, tuple<int, int, int>>> nums;
    vector<vector<char>> sym(a.size(), vector<char>(a[0].size(), 0));
    for (int j = 0; j < a.size(); ++j) {
        string num;
        for (int i = 0; i < a[j].length(); ++i) {
            if (isdigit(a[j][i])) {
                num += a[j][i];
                if (i + 1 == a[j].length() || !isdigit(a[j][i + 1])) {
                    nums.push_back({stoi(num), {j, i, num.size()}});
                    num = "";
                }
            }
            else if (a[j][i] != '.')
                sym[j][i] = a[j][i];
        }
    }

    task1(nums, sym);
    task2(nums, sym);
    return 0;
}